### ISSUE TITLE
Literal comparison for Decimal values

### ISSUE THREAD

**Issue by satra** [trungdong/prov#77]
@trungdong in prov/model.py the way literal comparison is implemented, it results in:

`pm.Literal(10, datatype=pm.XSD_DECIMAL) != pm.Literal(10.0, datatype=pm.XSD_DECIMAL)`

is that intended?

**Comment by trungdong** [comment]
@satra, this is unintended. It is caused by the conversion from the value of a `Literal` into string below:
```python
class Literal(object):
    def __init__(self, value, datatype=None, langtag=None):
        self._value = six.text_type(value)  # value is always a string
```
I'm not sure there is a good solution for this, unless the equality check of `Literal` understands XSD values.

This is clearly undesirable. However, I'm reluctant to implement the handling of XSD types in equality check, which might be time-consuming.

I wonder if there is a simple solution.

**Comment by satra** [comment]
i agree that this shouldn't be done in equality. by representing the value as a string, it may also have some unfortunate side effects especially for floating point numbers.

one option is to handle it in the representation as is done in rdflib, which maps literals to native python types.

http://rdflib.readthedocs.org/en/stable/rdf_terms.html#literals

a number of the current failures are from lexical comparison of decimals.

**Comment by trungdong** [comment]
Thank you for the pointer, @satra. We've actually done that for a few XSD datatypes, but now I can extend that to cover more with the example in rdflib.

**Comment by trungdong** [comment]
Confirmed, and still the case on current master: `Literal` stores its value as a string and compares lexically, so `Literal(10, XSD_DECIMAL)` and `Literal(10.0, XSD_DECIMAL)` compare unequal even though they denote the same xsd:decimal value. The fix is to compare (and parse) decimal literals in value space, but that changes equality/hashing — and therefore attribute deduplication, record equality, and round-trip results — for inputs that work today, so it is scheduled for the 3.0 release alongside the PROV-CONSTRAINTS unification rework rather than a 2.x patch. See [ROADMAP.md](https://github.com/trungdong/prov/blob/master/ROADMAP.md) and the roadmap tracking issue #181.
